### PR TITLE
OKTA-299521: migration guide to 3.0

### PR DESCRIPTION
**Pull request by bryanapellanes-okta**
This change adds information about breaking changes and additions in the upcoming 3.0 release of this SDK.

See preview here: https://github.com/okta/okta-sdk-dotnet/blob/OKTA-299521-migration-guide/MIGRATING.md

### CODE REVIEW

File: MIGRATING.md [hunk 1]
```diff
@@ -2,6 +2,94 @@
 
 This library uses semantic versioning and follows Okta's [library version policy](https://developer.okta.com/code/library-versions/). In short, we don't make breaking changes unless the major version changes!
 
+## Migrating to 3.0
```
Comment (laura-rodriguez) [inline MIGRATING.md:5]:
```suggestion
## Migrating to from 2.0.0 to 3.x
```

File: MIGRATING.md [hunk 2]
```diff
@@ -2,6 +2,94 @@
...
+
+Version 3.0 of this library introduces a number of breaking changes from previous versions; in addition to new classes some class definitions are no longer backward compatible due to method renames and signature changes, see [Breaking Changes](#breaking-changes).
```
Comment (laura-rodriguez) [inline MIGRATING.md:7]: Nit: Let's reference the project's version using Major.Minor.Patch  (3.0.0)

File: MIGRATING.md [hunk 3]
```diff
@@ -2,6 +2,94 @@
...
+### Breaking Changes
+
+The following is a list of changes that break backward compatibility in version 3.0.
+
+**Okta.Sdk.OktaClient**
+- `CreatedScoped(Okta.Sdk.RequestContext requestContext)`
+<br />&mdash; Renamed `CreateScoped(Okta.Sdk.RequestContext requestContext)`
+
+**Okta.Sdk.GroupsClient**                
+- `ListGroups(String q,String filter,String after,Int32 limit,String expand)`
```
Comment (laura-rodriguez) [inline MIGRATING.md:18]: Question: Are we using `String` or `string`?

File: MIGRATING.md [hunk 4]
```diff
@@ -2,6 +2,94 @@
...
+
+The following is a list of changes that break backward compatibility in version 3.0.
+
+**Okta.Sdk.OktaClient**
+- `CreatedScoped(Okta.Sdk.RequestContext requestContext)`
+<br />&mdash; Renamed `CreateScoped(Okta.Sdk.RequestContext requestContext)`
+
+**Okta.Sdk.GroupsClient**                
+- `ListGroups(String q,String filter,String after,Int32 limit,String expand)` 
+<br />&mdash; Signature changed `ListGroups(String q,String filter,String after,Int32 limit)`
```
Comment (laura-rodriguez) [inline MIGRATING.md:19]: Question: Are we using `Int32` or `int`?

File: MIGRATING.md [hunk 5]
```diff
@@ -2,6 +2,94 @@
...
+<br />&mdash; Signature changed `GetGroupAsync(String groupId,CancellationToken cancellationToken)`
+- `ListGroupUsers(String groupId,String after,Int32 limit,String managedBy)`
+<br />&mdash; Signature changed `ListGroupUsers(String groupId,String after,Int32 limit)`
+- `RemoveGroupUserAsync(String groupId,String userId,CancellationToken cancellationToken)`
+<br />&mdash; Renamed `RemoveUserFromGroupAsync(String groupId,String userId,CancellationToken cancellationToken)`
+
+**Okta.Sdk.PoliciesClient**
+- `ListPolicies(String type,String status,String after,Int32 limit,String expand)`
+<br />&mdash; Signature changed `ListPolicies(String type,String status,String expand)`
+- `AddPolicyRuleAsync(Okta.Sdk.IPolicyRule policyRule,String policyId,Boolean activate,CancellationToken cancellationToken)`
```
Comment (laura-rodriguez) [inline MIGRATING.md:44]: Question: Should we replace `Boolean` by `bool`?

File: MIGRATING.md [hunk 6]
```diff
@@ -2,6 +2,94 @@
...
+<br />&mdash; Signature changed `AddPolicyRuleAsync(Okta.Sdk.IPolicyRule policyRule,String policyId,CancellationToken cancellationToken)`
+
+**Okta.Sdk.UserFactorsClient**                
+- `AddFactorAsync(Okta.Sdk.IFactor factor,String userId,Boolean updatePhone,String templateId,Int32 tokenLifetimeSeconds,Boolean activate,CancellationToken cancellationToken)`
```
Comment (laura-rodriguez) [inline MIGRATING.md:48]: Nit: Consider a space after commas.

File: MIGRATING.md [hunk 7]
```diff
@@ -2,6 +2,94 @@
...
+  - `ExpirePasswordAndGetTemporaryPasswordAsync(String userId,CancellationToken cancellationToken)`
+- `ListAssignedRoles(String userId,String expand)`
+<br />&mdash; Renamed `ListAssignedRolesForUser(String userId,String expand)`
+- `EndAllUserSessionsAsync(String userId,Boolean oauthTokens,CancellationToken cancellationToken)`
+<br />&mdash; Renamed `ClearUserSessionsAsync(String userId,Boolean oauthTokens,CancellationToken cancellationToken)`
+
+### New Okta Clients
+The following is a list of context specific clients that are new in version 3.0.  Instances of each are available as properties of an OktaClient instance where the name of the property is the name of the type with the "Client" suffix removed.
+
+- `Okta.Sdk.AuthorizationServersClient`, see [Authorization Servers](https://developer.okta.com/docs/reference/api/authorization-servers/).
```
Comment (laura-rodriguez) [inline MIGRATING.md:81]:
```suggestion
- `Okta.Sdk.AuthorizationServersClient`, see [Authorization Servers API](https://developer.okta.com/docs/reference/api/authorization-servers/).
```

File: MIGRATING.md [hunk 8]
```diff
@@ -2,6 +2,94 @@
...
+- `Okta.Sdk.EventHooksClient`, see [Event Hooks Management API](https://developer.okta.com/docs/reference/api/event-hooks/).
+- `Okta.Sdk.FeaturesClient`, see [Features API](https://developer.okta.com/docs/reference/api/features/).
+- `Okta.Sdk.IdentityProvidersClient`, see [Identity Providers API](https://developer.okta.com/docs/reference/api/idps/).
+- `Okta.Sdk.InlineHooksClient`, see [Inline Hooks Management API](https://developer.okta.com/docs/reference/api/inline-hooks/).
+- `Okta.Sdk.LinkedObjectsClient`, see [Linked Objects API](https://developer.okta.com/docs/reference/api/linked-objects/).
+- `Okta.Sdk.TemplatesClient`, see [Custom Templates API](https://developer.okta.com/docs/reference/api/templates/).
+- `Okta.Sdk.TrustedOriginsClient`, see [Trusted Origins API](https://developer.okta.com/docs/reference/api/trusted-origins/).
```
Comment (laura-rodriguez) [inline MIGRATING.md:88]: Question: Did we add a client for this? I don't remember seeing this in the spec.

File: MIGRATING.md (hunk 1 above)
Comment (bryanapellanes-okta) [reply MIGRATING.md:5]: I'll change it.

File: MIGRATING.md (hunk 3 above)
Comment (bryanapellanes-okta) [reply MIGRATING.md:18]: These were generated using reflection; I'll change it.

File: MIGRATING.md (hunk 4 above)
Comment (bryanapellanes-okta) [reply MIGRATING.md:19]: Same here, reflection; I'll update.

File: MIGRATING.md (hunk 5 above)
Comment (bryanapellanes-okta) [reply MIGRATING.md:44]: Reflection; I'll update.

File: MIGRATING.md (hunk 6 above)
Comment (bryanapellanes-okta) [reply MIGRATING.md:48]: 👍

File: MIGRATING.md (hunk 7 above)
Comment (bryanapellanes-okta) [reply MIGRATING.md:81]: I lean toward agreeing with you.  However, the title here should match the title in the target.  The title at the target url (https://developer.okta.com/docs/reference/api/authorization-servers/) is "Authorization Servers".  All the others are suffixed by "API".  For consistency it should be changed in both places; I'll change it here and submit a PR in the docs repo.

File: MIGRATING.md (hunk 8 above)
Comment (bryanapellanes-okta) [reply MIGRATING.md:88]: Looks like this was added in the previous round of changes.

File: MIGRATING.md (hunk 7 above)
Comment (bryanapellanes-okta) [reply MIGRATING.md]: Updated here and submitted PR to update docs: https://github.com/okta/okta-developer-docs/pull/1277

File: MIGRATING.md [hunk 9]
```diff
@@ -2,6 +2,94 @@
...
+  - `ForgotPasswordSetNewPasswordAsync(Okta.Sdk.IUserCredentials user, string userId, bool sendEmail, CancellationToken cancellationToken)`
+  - `ExpirePasswordAsync(string userId, CancellationToken cancellationToken)`
+  - `ExpirePasswordAndGetTemporaryPasswordAsync(string userId, CancellationToken cancellationToken)`
+- `ListAssignedRoles(string userId, string expand)`
+<br />&mdash; Renamed `ListAssignedRolesForUser(string userId, string expand)`
+- `EndAllUserSessionsAsync(string userId, bool oauthTokens, CancellationToken cancellationToken)`
+<br />&mdash; Renamed `ClearUserSessionsAsync(string userId, bool oauthTokens, CancellationToken cancellationToken)`
+
+### New Okta Clients
+The following is a list of context specific clients that are new in version 3.0.  Instances of each are available as properties of an OktaClient instance where the name of the property is the name of the type with the "Client" suffix removed.
```
Comment (laura-rodriguez) [inline MIGRATING.md:79]:
```suggestion
The following is a list of context specific clients that are new in version 3.0.0. Instances of each are available as properties of an OktaClient instance where the name of the property is the name of the type with the "Client" suffix removed.
```

File: MIGRATING.md (hunk 8 above)
Comment (laura-rodriguez) [reply MIGRATING.md:88]: I see, but it is gonna be available in 3.0.0 right?

Comment (bryanapellanes-okta) [reply MIGRATING.md:88]: Yes

**Comment by bryanapellanes-okta** [comment]
@laura-rodriguez Accepted latest suggested change, answered question re: TrustedOriginsClient; yes it will be in version 3.0.0.  Merging pr now.

**Pull request merged**